### PR TITLE
Unify DF and CF result

### PR DESCRIPTION
**Pull request by gustavoguichard**

### CODE REVIEW

File: src/constructor.ts [hunk 1]
```diff
@@ -86,18 +47,20 @@ function fromComposable<I, E, A extends Composable>(
...
+        : envResult.error.issues.map(
+            (error) =>
+              new EnvironmentError(error.message, error.path as string[]),
+          )
+      return failure([...inputErrors, ...envErrors])
     }
-    return dfResultFromcomposable(fn)(
-      ...([result.data as I, envResult.data as E] as Parameters<A>),
-    )
+    return fn(result.data, envResult.data)
```
Comment (diogob) [inline src/constructor.ts:63]: this is the elegance we were striving for!
DFs are composables with an enviroment constant on composition + unknown inputs and runtime parsers

File: src/constructor.ts [hunk 2]
```diff
@@ -127,10 +90,9 @@ const undefinedSchema: ParserSchema<undefined> = {
...
 
 export {
-  dfResultFromcomposable,
   fromComposable,
-  makeDomainFunction,
   makeDomainFunction as mdf,
-  toComposable,
-  makeSuccessResult,
+  makeDomainFunction,
+  success as makeSuccessResult,
```
Comment (diogob) [inline src/constructor.ts]: we can probably remove

File: src/domain-functions.ts [hunk 3]
```diff
@@ -37,7 +36,7 @@ import { Composable } from './index.ts'
  * }
  */
 function safeResult<T>(fn: () => T): Promise<Result<T>> {
-  return dfResultFromcomposable(A.composable(fn))() as Promise<Result<T>>
+  return A.composable(fn)() as Promise<Result<T>>
 }
```
Comment (diogob) [inline src/domain-functions.ts]: this was not in the readme, perhaps is safe to remove

File: src/domain-functions.ts [hunk 4]
```diff
@@ -70,12 +75,10 @@ function applyEnvironment<
...
-  return ((input, environment) => {
-    const composables = fns.map((df) =>
-      A.composable(() => fromSuccess(df)(input, environment)),
-    )
-    return dfResultFromcomposable(A.all(...(composables as [Composable])))()
-  }) as DomainFunction<UnpackAll<Fns>>
+  return ((input, environment) =>
+    A.all(...applyEnvironmentToList(fns, environment))(
+      input,
+    )) as DomainFunction<UnpackAll<Fns>>
```
Comment (diogob) [inline src/domain-functions.ts:55]: :smiling_face_with_tear:

File: src/domain-functions.ts [hunk 5]
```diff
@@ -235,9 +228,7 @@ function map<O, R>(
   dfn: DomainFunction<O>,
   mapper: (element: O) => R | Promise<R>,
 ): DomainFunction<R> {
-  return dfResultFromcomposable(
-    A.map(A.composable(fromSuccess(dfn)), mapper),
-  ) as DomainFunction<R>
+  return A.map(dfn, mapper)
```
Comment (diogob) [inline src/domain-functions.ts:205]: This is so cool, I tried just assigning a typed const and it also works perfectly. But I like having the named function on the top level of the module.

File: src/index.ts [hunk 6]
```diff
@@ -1,31 +1,54 @@
...
-  toComposable,
   makeSuccessResult,
+  mdf,
+  success,
 } from './constructor.ts'
-export * from './domain-functions.ts'
-export * from './input-resolvers.ts'
-export * from './errors.ts'
+export {
+  all,
```
Comment (diogob) [inline src/index.ts:8]: yay, no more surprise export! :tada:

File: src/serialize.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import { EnvironmentError } from './errors.ts'
+import { InputError } from './errors.ts'
+import type { Result } from './types.ts'
+import { Failure } from './types.ts'
+import { Success } from './types.ts'
+
+function serializeResult<T>(result: Result<T>):
```
Comment (diogob) [inline src/serialize.ts]: it seems we have some additional work to be done here, I suggest we merge as is and create PR just for backward compatibility at the end of our work. Just so we travel light.

File: src/constructor.ts (hunk 1 above)
Comment (gustavoguichard) [reply src/constructor.ts:63]: Took note of that definition!

File: src/domain-functions.ts (hunk 3 above)
Comment (gustavoguichard) [reply src/domain-functions.ts]: Let's!

File: src/domain-functions.ts (hunk 5 above)
Comment (gustavoguichard) [reply src/domain-functions.ts:205]: I thought about the same... would like to see how you've done it

File: src/serialize.ts (hunk 7 above)
Comment (gustavoguichard) [reply src/serialize.ts]: Yup, I'll remove it from this PR

File: src/domain-functions.ts [hunk 8]
```diff
@@ -167,8 +138,8 @@ function merge<Fns extends DomainFunction<Record<string, unknown>>[]>(
 function pipe<T extends DomainFunction[]>(
   ...fns: T
 ): DomainFunction<Last<UnpackAll<T>>> {
-  const last = <T>(ls: T[]): T => ls[ls.length - 1]
-  return map(sequence(...fns), last) as DomainFunction<Last<UnpackAll<T>>>
```
Comment (gustavoguichard) [inline src/domain-functions.ts:171]: @diogob let's take a note from this implementation to simplify the composable's `pipe`

File: src/domain-functions.ts (hunk 5 above)
Comment (diogob) [reply src/domain-functions.ts:205]:
```typescript
const map: <O, R>(
  dfn: DomainFunction<O>,
  mapper: (element: O) => R | Promise<R>,
) => DomainFunction<R> = A.map
```

File: src/errors.ts [hunk 9]
```diff
@@ -63,119 +28,49 @@ class InputErrors extends Error {
...
+ * @returns the Error
+ * @example
+ * try {}
+ * catch (error) {
+ *   const Error = toError(error)
+ *   console.log(Error.message)
+ * }
+ */
+function toError(maybeError: unknown): Error {
+  return isError(maybeError) ? maybeError : new Error(String(maybeError))
```
Comment (gustavoguichard) [inline src/errors.ts]: @diogob I'd simplify this like so:
```suggestion
  return maybeError instanceof Error ? maybeError : new Error(JSON.stringify(maybeError))
```

But it breaks one test bc it wouldn't allow:
```ts
throw { message: 'Foo' }
```

Which is a very weird API TBH

**Pull request merged**